krenew.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.19)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "KRENEW 1"
.TH KRENEW 1 "2012-01-07" "4.1" "kstart"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
krenew \- Renew a Kerberos ticket
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBkrenew\fR [\fB\-bhiLstvx\fR] [\fB\-c\fR \fIchild pid file\fR] [\fB\-H\fR \fIminutes\fR]
    [\fB\-K\fR \fIminutes\fR] [\fB\-k\fR \fIticket cache\fR] [\fB\-p\fR \fIpid file\fR]
    [\fIcommand\fR ...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBkrenew\fR renews an existing renewable ticket.  When run without any
arguments, it just attempts to renew the existing ticket-granting ticket
in the current ticket cache, equivalent to \f(CW\*(C`kinit \-R\*(C'\fR, but it can
optionally run a program like \fBaklog\fR to refresh \s-1AFS\s0 tokens, can run as a
daemon and wake up periodically to renew the ticket cache, or can run a
specified command and keep renewing the ticket cache until the command
finishes (or renewal is no longer possible).  If a command is specified,
\&\fBkrenew\fR by default wakes up every 60 minutes (1 hour) to check the
ticket cache.
.PP
If a command is given, \fBkrenew\fR makes a copy of the ticket cache and
creates a private ticket cache just for that command, thus isolating it
from later destruction of the original ticket cache.  This allows krenew
to maintain authentication for a command even if, for example, the user
running the command logs out and OpenSSH destroys their original ticket
cache.
.PP
If a command is given, it will not be run using the shell, so if you want
to use shell metacharacters in the command with their special meaning,
give \f(CW\*(C`sh \-c \f(CIcommand\f(CW\*(C'\fR as the command to run and quote \fIcommand\fR.  If
the command contains command-line options (like \f(CW\*(C`\-c\*(C'\fR), put \f(CW\*(C`\-\-\*(C'\fR on the
command line before the beginning of the command to tell \fBkrenew\fR to not
parse those options as its own.
.PP
If \fBkrenew\fR is built with \fIsetpag()\fR support and \s-1AFS\s0 tokens are requested
with the \fB\-t\fR option, it will put the command in a separate \s-1PAG\s0 before
obtaining \s-1AFS\s0 tokens so that they don't interfere with other processes on
the system.
.PP
When running a command, \fBkrenew\fR propagates \s-1HUP\s0, \s-1TERM\s0, \s-1INT\s0, and \s-1QUIT\s0
signals to the child process and does not exit when those signals are
received.  (If the propagated signal causes the child process to exit,
\&\fBkrenew\fR will then exit.)  This allows \fBkrenew\fR to react properly when
run under a command supervision system such as \fIrunit\fR\|(8) or \fIsvscan\fR\|(8) that
uses signals to control supervised commands, and to run interactive
commands that should receive Ctrl-C.
.PP
If a running \fBkrenew\fR receives an \s-1ALRM\s0 signal, it immediately refreshes
the ticket cache regardless of whether it is in danger of expiring.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-b\fR" 4
.IX Item "-b"
After starting, detach from the controlling terminal and run in the
background.  This option only makes sense in combination with \fB\-K\fR or a
command that \fBkrenew\fR will be running.  \fBkrenew\fR will not background
itself until after it does the initial ticket renewal, so that any initial
errors will be reported, but it will then redirect output to \fI/dev/null\fR
and no subsequent errors will be reported.
.Sp
If this flag is given, \fBkrenew\fR will also change directories to \f(CW\*(C`/\*(C'\fR.
All paths (such as to a command to run or a \s-1PID\s0 file) should therefore be
given as absolute, not relative, paths.
.Sp
If used in conjunction with a command to run, that command will also run
in the background and will also have its input and output redirected to
\&\fI/dev/null\fR.  It will have to report any errors via some other mechanism
for the errors to be seen.
.Sp
Use of this flag on Mac \s-1OS\s0 X without specifying a file-based ticket cache
by either using \fB\-k\fR or setting \s-1KRB5CCNAME\s0 will probably not do what you
want.  Ticket caches on Mac \s-1OS\s0 X are, by default, per-session and with
\&\fB\-b\fR \fBkrenew\fR will detach itself from your existing ticket cache.
Instead, to renew the default ticket cache on Mac \s-1OS\s0 X, try something
like:
.Sp
.Vb 1
\&    (krenew \-K 60 &)
.Ve
.Sp
to run \fBkrenew\fR in the background but within the current session.
.Sp
When using this option, consider also using \fB\-L\fR to report \fBkrenew\fR
errors to syslog.
.IP "\fB\-c\fR \fIchild pid file\fR" 4
.IX Item "-c child pid file"
Save the process \s-1ID\s0 (\s-1PID\s0) of the child process into \fIchild pid file\fR.
\&\fIchild pid file\fR is created if it doesn't exist and overwritten if it
does exist.  This option is only allowed when a command was given on the
command line and is most useful in conjunction with \fB\-b\fR to allow
management of the running child process.
.Sp
Note that, when used with \fB\-b\fR, the \s-1PID\s0 file is written out after
\&\fBkrenew\fR is backgrounded and changes its working directory to \fI/\fR, so
relative paths for the \s-1PID\s0 file will be relative to \fI/\fR (probably not
what you want).
.IP "\fB\-H\fR \fIminutes\fR" 4
.IX Item "-H minutes"
Only renew the ticket if it has a remaining lifetime of less than
\&\fIminutes\fR minutes.  If either the ticket already has a sufficiently long
remaining lifetime or renewal was successful, run the command (if one was
specified) or exit immediately with status 0 (if none was).  Otherwise,
try to renew the ticket so that it will have a remaining lifetime of at
least \fIminutes\fR, exit with an error if unsuccessful, and then run the
command, if any.  Cannot be used with \fB\-K\fR.
.IP "\fB\-h\fR" 4
.IX Item "-h"
Display a usage message and exit.
.IP "\fB\-i\fR" 4
.IX Item "-i"
Ignore errors in renewing the ticket and keep running.  Normally,
\&\fBkrenew\fR exits as soon as the ticket cache either disappears or the
tickets run out of renewable lifetime.  If this flag is given, it will
complain about the failure to standard error (unless \fB\-b\fR was given) but
continue running, waking up to try again after the next check interval
(see \fB\-K\fR).  This is useful if some other process may recreate an expired
ticket cache and \fBkrenew\fR should stay around and act on that recreated
ticket cache once it's present.
.Sp
Setup errors opening the Kerberos ticket cache or running the command will
still cause \fBkrenew\fR to exit, even if this flag is given.  Only Kerberos
errors and errors after \fBkrenew\fR is running and any command has been
started will be ignored.
.Sp
This flag is only useful in daemon mode or when a command was given.
.IP "\fB\-K\fR \fIminutes\fR" 4
.IX Item "-K minutes"
Run in daemon mode to keep a ticket alive indefinitely.  The program
reawakens after \fIminutes\fR minutes, checks if the ticket will expire
before or less than two minutes before the next scheduled check, and
renews the ticket if needed.  If this option is not given but a command
was given on the command line, the default interval is 60 minutes (1
hour).
.Sp
If an error occurs in refreshing the ticket cache that doesn't cause
\&\fBkrenew\fR to exit, the wake-up interval will be shortened to one minute
and the operation retried at that interval for as long as the error
persists.
.IP "\fB\-k\fR \fIticket cache\fR" 4
.IX Item "-k ticket cache"
Use \fIticket cache\fR as the ticket cache rather than the contents of the
environment variable \s-1KRB5CCNAME\s0 or the library default.  \fIticket cache\fR
may be any ticket cache identifier recognized by the underlying Kerberos
libraries.  This generally supports a path to a file, with or without a
leading \f(CW\*(C`FILE:\*(C'\fR string, but may also support other ticket cache types.
.IP "\fB\-L\fR" 4
.IX Item "-L"
Report messages to syslog as well as to standard output or standard error.
All messages will be logged with facility \s-1LOG_DAEMON\s0.  Regular messages
that are displayed on standard output are logged with level \s-1LOG_NOTICE\s0.
Errors that don't cause \fBkrenew\fR to terminate when run with \fB\-i\fR are
logged with level \s-1LOG_WARNING\s0.  Fatal errors are logged with level
\&\s-1LOG_ERR\s0.
.Sp
This is useful when debugging problems in combination with \fB\-b\fR.
.IP "\fB\-p\fR \fIpid file\fR" 4
.IX Item "-p pid file"
Save the process \s-1ID\s0 (\s-1PID\s0) of the running \fBkrenew\fR process into \fIpid
file\fR.  \fIpid file\fR is created if it doesn't exist and overwritten if it
does exist.  This option is most useful in conjunction with \fB\-b\fR to allow
management of the running \fBkrenew\fR daemon.
.Sp
Note that, when used with \fB\-b\fR the \s-1PID\s0 file is written out after
\&\fBkrenew\fR is backgrounded and changes its working directory to \fI/\fR, so
relative paths for the \s-1PID\s0 file will be relative to \fI/\fR (probably not
what you want).
.IP "\fB\-s\fR" 4
.IX Item "-s"
Normally, when \fBkrenew\fR exits abnormally while running a command (if, for
example, the ticket's renewable lifetime has expired), it leaves the
command running.  If \fB\-s\fR is given, it will send a \s-1SIGHUP\s0 signal to the
command before exiting.  This can be useful if it's pointless for the
command to keep running without Kerberos tickets.
.IP "\fB\-t\fR" 4
.IX Item "-t"
Run an external program after getting a ticket.  The default use of this
is to run \fBaklog\fR to get a token.  If the environment variable \s-1KINIT_PROG\s0
is set, it overrides the compiled-in default.
.Sp
If \fBkrenew\fR has been built with \s-1AFS\s0 \fIsetpag()\fR support and a command was
given on the command line, \fBkrenew\fR will create a new \s-1PAG\s0 before
obtaining \s-1AFS\s0 tokens.  Otherwise, it will obtain tokens in the current
\&\s-1PAG\s0.
.IP "\fB\-v\fR" 4
.IX Item "-v"
Be verbose.  This will print out a bit of additional information about
what is being attempted and what the results are.
.IP "\fB\-x\fR" 4
.IX Item "-x"
Exit immediately on any error.  Normally, when running a command or when
run with the \fB\-K\fR option, \fBkrenew\fR keeps running even if it fails to
renew the ticket cache as long as the ticket cache still exists and
appears to be renewable.  It tries again at the next check interval.  With
this option, \fBkrenew\fR will instead exit.
.SH "RETURN VALUES"
.IX Header "RETURN VALUES"
The program normally exits with status 0 if it successfully renews a
ticket.  If \fBkrenew\fR runs aklog or some other program \fBkrenew\fR returns
the exit status of that program.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Renew the current ticket-granting ticket.
.PP
.Vb 1
\&    krenew
.Ve
.PP
Wake up every ten minutes and check to see if the ticket cache needs
renewing.  If it does, re-run \fBaklog\fR as well.
.PP
.Vb 1
\&    krenew \-K 10 \-t
.Ve
.PP
Run the program \fI/usr/local/bin/compute\-job\fR in the background, checking
every hour to see if the ticket needs to be renewed (the default).  Put
the \s-1PID\s0 of the \fBkrenew\fR job in \fI/var/run/compute.pid\fR.  Obtain a new \s-1AFS\s0
token each time the ticket has to be renewed.
.PP
.Vb 1
\&    krenew \-b \-t \-p /var/run/compute.pid /usr/local/bin/compute\-job
.Ve
.PP
If you wanted to pass options to \fI/usr/local/bin/compute\-job\fR, putting a
\&\f(CW\*(C`\-\-\*(C'\fR argument before it would be necessary to keep \fBkrenew\fR from
interpreting those options as its own.
.PP
If you want to redirect output to a file that requires authentication to
write to, you will need to do that redirection in a sub-shell.  In other
words, the following command:
.PP
.Vb 1
\&    krenew \-t compute\-job > /afs/local/data/output
.Ve
.PP
won't work if /afs/local/data/output requires an \s-1AFS\s0 token to write to.
The job, while running, will have an \s-1AFS\s0 token, but the output redirection
is done in the parent shell and doesn't benefit from \fBkrenew\fR.  The above
should instead be written as:
.PP
.Vb 1
\&    krenew \-t \-\- sh \-c \*(Aqcompute\-job > /afs/local/data/output\*(Aq
.Ve
.PP

NEWS
                       User-Visible kstart Changes

kstart 4.1 (2012-01-07)

    Fix a regression introduced in kstart 4.0 that caused k5start -H and
    krenew -H to fail and attempt reauthentication with non-renewable
    tickets even if the lifetime was long enough.  Thanks to pod for the
    report.

    Fix a regression introduced in kstart 4.0 where k5start -H would be
    happy with an unexpired ticket for a different principal than the
    desired client principal.

    When k5start or krenew are running as a daemon and obtaining new
    tickets fails, both now shorten the wake-up interval to one minute and
    keep trying at that interval until the error resolves itself, and then
    go back to the normal wakeup interval.

    Add a new -s option to krenew that, if given, tells krenew to send
    SIGHUP to the command it's running when it exits because it can't
    renew the ticket.  This is useful when continuing to run the command
    without a valid ticket would be pointless.

    After a SIGHUP or SIGTERM when not running a command, k5start and
    krenew now clean up their PID files, if any, before exiting.

kstart 4.0 (2011-12-29)

    Remove k4start from the distribution.  I no longer have a Kerberos v4
    environment with which to test and therefore no way to refactor and
    restructure the code for other changes to the package.  Users who
    still need k4start should use an older version of the package.

    Ticket caches passed to k5start or krenew with the -k option are now
    used as-is without prepending "FILE:".  This allows both programs to
    be used with non-file caches (unles the -o, -g, or -m options were
    given to k5start, of course).  However, users who were relying on
    k5start or krenew prepending "FILE:" may now need to add this
    explicitly to the -k argument if they want the ticket cache to be set
    in the environment with that prefix.

    Always canonicalize the ticket cache name in k5start before
    propagating KRB5CCNAME to child processes.  This combined with the
    previous change allows -k to specify a ticket cache name that changes
    once the cache is created, such as when creating new PIPE caches.

    krenew now defaults to staying running if renewing credentials fails.
    The new -x option restores the previous behavior of exiting on any
    error.  It will still exit by default (unless -i is used) if the
    renewable lifetime has expired or if the ticket cache has been
    removed.

    k5start no longer exits on failure to obtain credentials when running
    as a daemon.  The new -x option restores the previous behavior of
    exiting on any error.  It does still exit if the first attempt to
    obtain credentials during startup (before backgrounding) fails, to
    make it easier to diagnose configuration errors.

    k5start, when run with the -o, -g, or -m options to change ticket
    cache ownership or permissions, now writes a temporary ticket cache in
    the same directory, sets its ownership and permissions, and then
    replaces the existing cache with an atomic rename.  It also sets
    permissions properly if it has to reauthenticate after backgrounding
    itself.  This closes two windows where the cache may not be accessible
    to the program using it if k5start were in the middle of refreshing
    it.  Thanks to Harry Coin for the report.

    k5start and krenew now propagate SIGINT (Ctrl-C) to the child process
    when running a command rather than exiting immediately.

    Set signal handlers with sigaction instead of signal, which may fix
    problems propagating multiple signals to child processes in k5start
    and krenew.

    Diagnose the nonsensical combination of -U and -u or -i options in
    k5start and report an error rather than ignoring -u and appending the
    instance from -i onto the principal obtained via -U.  Also diagnose
    the non-sensical combination of -H and a command to run in both
    k5start and krenew; just omit the -H flag for this case.

    Update the included kafs library to the version from rra-c-util 4.0,
    adding support for Mac OS X and Solaris 11.

    Change references to Kerberos v5 to just Kerberos in the
    documentation.  Kerberos v5 has been the default version of Kerberos
    for over ten years now.

    Update to rra-c-util 4.0:

    * Build on systems where krb5/krb5.h exists but krb5.h does not.
    * Build with OpenBSD Heimdal where there is no separate roken library.
    * Kerberos probes no longer assume transitive library dependencies.
    * Fix removal of /usr/include from Kerberos CPPFLAGS.
    * Add notices to all files copied from rra-c-util.
    * Fix replacement of krb5_free_error_message.
    * Support older Heimdal with no-context krb5_get_init_creds_opt_free.
    * Improve probe for krb5_kt_free_entry.
    * Fix use of long long, where available, in replacement mkstemp.
    * Include strings.h where present for more POSIX string functions.
    * Use typedef for a missing sig_atomic_t.
    * Avoid passing a NULL context to krb5_get_error_message.
    * Fix integer data types in the messages utility library.
    * Use configure-detected aklog path in the test suite.
    * Add replacement for a missing strndup (such as on Mac OS X).
    * Add tests for messages-krb5 utility functions.
    * Update compiler warning flags for make warnings to gcc 4.6.1.

    Update to C TAP Harness 1.9:

    * Add a usage message and -h option to runtests.
    * Honor -s and SOURCE in runtests even if BUILD is not set.
    * Improve test summary at the end of a C test case.
    * Flush stderr before printing TAP output.
    * Improve portability of output functions in the shell libtap.sh.
    * Add notices to all files copied from C TAP Harness.

kstart 3.16 (2010-01-19)

    Add the -L option to k5start and krenew, saying to log messages to
    syslog as well as standard output or standard error.


krenew.c
/*
 * Automatically renew a Kerberos ticket.
 *
 * Similar to k5start, krenew can run as a daemon or run a specified program
 * and wait until it completes.  Rather than obtaining fresh Kerberos
 * credentials, however, it uses an existing Kerberos ticket cache and tries
 * to renew the tickets until it exits or until the ticket cannot be renewed
 * any longer.
 *
 * Written by Russ Allbery <rra@stanford.edu>
 * Copyright 2006, 2007, 2008, 2009, 2010, 2011, 2012
 *     The Board of Trustees of the Leland Stanford Junior University
 *
 * See LICENSE for licensing terms.
 */

#include <config.h>
#include <portable/krb5.h>
#include <portable/system.h>

#include <signal.h>
#include <sys/stat.h>
#include <syslog.h>
#include <time.h>

#include <internal.h>
#include <util/concat.h>
#include <util/macros.h>
#include <util/messages.h>
#include <util/messages-krb5.h>
#include <util/xmalloc.h>

/* Holds the command-line options we need to pass to our callbacks. */
struct krenew_private {
    bool signal_child;          /* Kill child on abnormal exit. */
};

/* The usage message. */
const char usage_message[] = "\
Usage: krenew [options] [command]\n\
   -b                   Fork and run in the background\n\
   -c <file>            Write child process ID (PID) to <file>\n\
   -H <limit>           Check for a happy ticket, one that doesn't expire in\n\
                        less than <limit> minutes, and exit 0 if it's okay,\n\
                        otherwise renew the ticket\n\
   -h                   Display this usage message and exit\n\
   -i                   Keep running even if the ticket cache goes away or\n\
                        the ticket can no longer be renewed\n\
   -K <interval>        Run as daemon, renew ticket every <interval> minutes\n\
   -k <cache>           Use <cache> as the ticket cache\n\
   -L                   Log messages via syslog as well as stderr\n\
   -p <file>            Write process ID (PID) to <file>\n\
   -s                   Send SIGHUP to command when ticket cannot be renewed\n\
   -t                   Get AFS token via aklog or AKLOG\n\
   -v                   Verbose\n\
   -x                   Exit immediately on any error\n\
\n\
If the environment variable AKLOG (or KINIT_PROG for backward compatibility)\n\
is set to a program (such as aklog) then this program will be executed when\n\
requested by the -t flag.  Otherwise, %s.\n";

/* Included in the usage message if AFS support is compiled in. */
const char usage_message_kafs[] = "\n\
When invoked with -t and a command, krenew will create a new AFS PAG for\n\
the command before running the AKLOG program to keep its AFS credentials\n\
isolated from other processes.\n";


/*
 * Print out the usage message and then exit with the status given as the
 * only argument.  If status is zero, the message is printed to standard
 * output; otherwise, it is sent to standard error.
 */
static void
usage(int status)
{
    fprintf((status == 0) ? stdout : stderr, usage_message,
            ((PATH_AKLOG[0] == '\0')
             ? "using -t is an error"
             : "the program executed will be\n" PATH_AKLOG));
#ifdef HAVE_KAFS
    fprintf((status == 0) ? stdout : stderr, usage_message_kafs);
#endif
    exit(status);
}


/*
 * Given the Kerberos context and a pointer to the ticket cache, copy that
 * ticket cache to a new cache and return a newly allocated string for the
 * name of the cache.
 */
static char *
copy_cache(krb5_context ctx, krb5_ccache *ccache)
{
    krb5_error_code code;
    krb5_ccache old, new;
    krb5_principal princ = NULL;
    char *name;
    int fd;

    if (xasprintf(&name, "/tmp/krb5cc_%d_XXXXXX", (int) getuid()) < 0)
        die("cannot format ticket cache name");
    fd = mkstemp(name);
    if (fd < 0)
        sysdie("cannot create ticket cache file");
    if (fchmod(fd, 0600) < 0)
        sysdie("cannot chmod ticket cache file");
    code = krb5_cc_resolve(ctx, name, &new);
    if (code != 0)
        die_krb5(ctx, code, "error initializing new ticket cache");
    old = *ccache;
    code = krb5_cc_get_principal(ctx, old, &princ);
    if (code != 0)
        die_krb5(ctx, code, "error getting principal from old cache");
    code = krb5_cc_initialize(ctx, new, princ);
    if (code != 0)
        die_krb5(ctx, code, "error initializing new cache");
    krb5_free_principal(ctx, princ);
    code = krb5_cc_copy_cache(ctx, old, new);
    if (code != 0)
        die_krb5(ctx, code, "error copying credentials");
    code = krb5_cc_close(ctx, old);
    if (code != 0)
        die_krb5(ctx, code, "error closing old ticket cache");
    *ccache = new;
    return name;
}


/*
 * Renew the user's tickets, warning if this isn't possible.  This is the
 * callback passed to the generic framework.  Takes the context, the
 * configuration, and a status code.  For the first authentication or on
 * SIGALRM, the status code will be 0; for other authentications, the status
 * code will be whatever is returned by ticket_expired, and therefore will be
 * KRB5KRB_AP_ERR_TKT_EXPIRED if the ticket needs to be renewed.  If the code
 * is KRB5KDC_ERR_KEY_EXP, that means we cannot renew the ticket for long
 * enough.
 *
 * Returns a Kerberos error code, which the framework chooses whether or not
 * to ignore.
 */
static krb5_error_code
renew(krb5_context ctx, struct config *config, krb5_error_code status)
{
    krb5_ccache ccache = NULL;
    krb5_error_code code;
    krb5_principal user = NULL;
    krb5_creds creds;
    bool creds_valid = false;

    /*
     * If we can't read the cache, or if we can't renew tickets for long
     * enough, exit unless run with -i.
     */
    if (status != 0 && status != KRB5KRB_AP_ERR_TKT_EXPIRED) {
        if (status == KRB5KDC_ERR_KEY_EXP)
            warn("ticket cannot be renewed for long enough");
        else
            warn_krb5(ctx, status, "error reading ticket cache");
        if (!config->ignore_errors)
            exit_cleanup(ctx, config, 1);
        return status;
    }
    memset(&creds, 0, sizeof(creds));
    code = krb5_cc_resolve(ctx, config->cache, &ccache);
    if (code != 0) {
        warn_krb5(ctx, code, "error opening ticket cache");
        if (!config->ignore_errors)
            exit_cleanup(ctx, config, 1);
        return code;
    }
    code = krb5_cc_get_principal(ctx, ccache, &user);
    if (code != 0) {
        warn_krb5(ctx, code, "error reading ticket cache");
        krb5_cc_close(ctx, ccache);
        if (!config->ignore_errors)
            exit_cleanup(ctx, config, 1);
        return code;
    }
    if (config->verbose) {
        char *name;

        code = krb5_unparse_name(ctx, user, &name);
        if (code != 0)
            warn_krb5(ctx, code, "error unparsing name");
        else {
            notice("renewing credentials for %s", name);
            krb5_free_unparsed_name(ctx, name);
        }
    }

    /*
     * If we just can't renew or store tickets, we keep trying unless -x was
     * given, which means we return an error code and let the framework handle
     * it.
     */
    code = krb5_get_renewed_creds(ctx, &creds, user, ccache, NULL);
    creds_valid = true;
    if (code != 0) {
        warn_krb5(ctx, code, "error renewing credentials");
        goto done;
    }
    
    /*
     * In theory, we don't want to reinitialize the cache and instead want to
     * just store the new credentials.  By reinitializing the cache, we create
     * a window where the credentials aren't valid.  However, I don't know how
     * to just store the renewed credentials without creating a cache that
     * grows forever.
     */
    code = krb5_cc_initialize(ctx, ccache, user);
    if (code != 0) {
        warn_krb5(ctx, code, "error reinitializing cache");
        goto done;
    }
    code = krb5_cc_store_cred(ctx, ccache, &creds);
    if (code != 0) {
        warn_krb5(ctx, code, "error storing credentials");
        goto done;
    }

done:
    if (ccache != NULL)
        krb5_cc_close(ctx, ccache);
    if (user != NULL)
        krb5_free_principal(ctx, user);
    if (creds_valid)
        krb5_free_cred_contents(ctx, &creds);
    return code;
}


/*
 * The cleanup callback.  All that we do here is send SIGHUP to the child
 * process if it's still running (config->child isn't 0) and we were
 * configured to do so.
 */
static void
cleanup(krb5_context ctx UNUSED, struct config *config,
        krb5_error_code status UNUSED)
{
    if (config->child != 0 && config->private.krenew->signal_child)
        kill(config->child, SIGHUP);
}


int
main(int argc, char *argv[])
{
    int option;
    krb5_context ctx;
    krb5_error_code code;
    struct config config;
    struct krenew_private private;
    krb5_ccache ccache;

    /* Initialize logging. */
    message_program_name = "krenew";

    /* Set up configuration and parse command-line options. */
    memset(&config, 0, sizeof(config));
    memset(&private, 0, sizeof(private));
    config.private.krenew = &private;
    config.auth = renew;
    config.cleanup = cleanup;
    while ((option = getopt(argc, argv, "bc:H:hiK:k:Lp:qstvx")) != EOF)
        switch (option) {
        case 'b': config.background = true;     break;
        case 'c': config.childfile = optarg;    break;
        case 'h': usage(0);                     break;
        case 'i': config.ignore_errors = true;  break;
        case 'k': config.cache = optarg;        break;
        case 'p': config.pidfile = optarg;      break;
        case 's': private.signal_child = true;  break;
        case 't': config.do_aklog = true;       break;
        case 'v': config.verbose = true;        break;
        case 'x': config.exit_errors = true;    break;

        case 'H':
            config.happy_ticket = convert_number(optarg, 10);
            if (config.happy_ticket <= 0)
                die("-H limit argument %s invalid", optarg);
            break;
        case 'K':
            config.keep_ticket = convert_number(optarg, 10);
            if (config.keep_ticket <= 0)
                die("-K interval argument %s invalid", optarg);
            break;
        case 'L':
            openlog(message_program_name, LOG_PID, LOG_DAEMON);
            message_handlers_notice(2, message_log_stdout,
                                    message_log_syslog_notice);
            message_handlers_warn(2, message_log_stderr,
                                  message_log_syslog_warning);
            message_handlers_die(2, message_log_stderr,
                                 message_log_syslog_err);
            break;

        default:
            usage(1);
            break;
        }

    /* Parse arguments.  If any are given, they will be the command to run. */
    argc -= optind;
    argv += optind;
    if (argc > 0)
        config.command = argv;

    /* Check the arguments for consistency. */
    if (config.background && config.keep_ticket == 0 && config.command == NULL)
        die("-b only makes sense with -K or a command to run");
    if (config.happy_ticket > 0 && config.keep_ticket > 0)
        die("-H and -K options cannot be used at the same time");
    if (config.happy_ticket > 0 && config.command != NULL)
        die("-H option cannot be used with a command");
    if (config.childfile != NULL && config.command == NULL)
        die("-c option only makes sense with a command to run");
    if (private.signal_child && config.command == NULL)
        die("-s option only makes sense with a command to run");

    /* Establish a Kerberos context and set the ticket cache. */
    code = krb5_init_context(&ctx);
    if (code != 0)
        die_krb5(ctx, code, "error initializing Kerberos");
    if (config.cache == NULL)
        code = krb5_cc_default(ctx, &ccache);
    else
        code = krb5_cc_resolve(ctx, config.cache, &ccache);
    if (code != 0)
        die_krb5(ctx, code, "error opening default ticket cache");
    if (config.command != NULL) {
        config.cache = copy_cache(ctx, &ccache);
        config.clean_cache = true;
    }
    if (config.cache == NULL) {
        code = krb5_cc_get_full_name(ctx, ccache, (char **) &config.cache);
        if (code != 0)
            die_krb5(ctx, code, "error getting ticket cache name");
    } else {
        if (setenv("KRB5CCNAME", config.cache, 1) != 0)
            die("cannot set KRB5CCNAME environment variable");
    }
    krb5_cc_close(ctx, ccache);

    /* Do the actual work. */
    run_framework(ctx, &config);
}
